executor/src/witgen/machines/block_machine.rs

```diff
@@ -276,42 +241,31 @@ impl<T: FieldElement> BlockMachine<T> {
         {
             // All values on the left hand side are known, check if this is a query
             // to the last row.
-            let row_before = self.row;
-            self.row = self.rows() - 1;
-            let result = self.process_outer_query(fixed_data, left, right);
-
-            match result {
-                Ok(result) => {
-                    if result.is_complete() {
-                        return Ok(result);
-                    } else {
-                        // If the query is not complete, we roll back the row change and continue.
-                        // This might happen if the block machine just validates some input (rather
-                        // than computing an output) and hasn't been called on that input before.
-                        self.row = row_before;
-                    }
-                }
-                Err(e) => {
-                    // Non-recoverable error
-                    return Err(e);
-                }
+            let row = self.rows() - 1;
+
+            let current = &self.data[row as usize];
+            // We don't have the next row, because it would be the first row of the next block.
+            // We'll use a fresh row instead.
+            let next = self.row_factory.fresh_row();
+            let row_pair = RowPair::new(current, &next, row, fixed_data, false);
+
+            let result = identity_processor.process_backlink(left, right, &row_pair)?;
```

Comment:
Can you explain why `process_backlink` is needed? Why doesn't it happen as part of the external call, right after the callee returns?

Reply:
I renamed it to `process_link`, as it goes both ways!

Reply:
I understand the forward way but not really the backward one:

In your example:
```
caller calls process_link[1, 2, c] on callee
callee returns {c: 3}
caller updates c to 3
```

If you use this method both ways, does that mean that you execute `callee calls process_link[1, 2, 3] on caller` instead of the last two steps above?

Reply:
I'd say it's more like:
- Caller calls `process_plookup()` on the machine
- Callee instantiates an `IdentityProcessor` and repeatedly calls `process_link` on it to update it's internal state, which includes a mutable copy of the relevant cells of the callee (`c` in this example). For these cells, all updates are both applied locally and also recorded.
- The caller returns the recorded updates
- The Callee applies them